services/webnn/ort/graph_builder_ort.cc

```diff
@@ -505,13 +506,161 @@ void GraphBuilderOrt::FindBoolOperands() {
+        break;
+      }
+      case mojom::Operation::Tag::kReshape: {
+        const auto& reshape = *operation->get_reshape();
+        uint8_input_operands.insert(reshape.input_operand_id);
+        break;
+      }
+      case mojom::Operation::Tag::kReverse: {
+        const auto& reverse = *operation->get_reverse();
+        uint8_input_operands.insert(reverse.input_operand_id);
+        break;
+      }
+      case mojom::Operation::Tag::kScatterElements: {
+        const auto& scatter_elements = *operation->get_scatter_elements();
+        uint8_input_operands.insert(scatter_elements.input_operand_id);
+        break;
+      }
+      case mojom::Operation::Tag::kScatterNd: {
+        const auto& scatter_nd = *operation->get_scatter_nd();
+        uint8_input_operands.insert(scatter_nd.input_operand_id);
+        uint8_input_operands.insert(scatter_nd.updates_operand_id);
+        break;
+      }
+      case mojom::Operation::Tag::kSlice: {
+        const auto& slice = *operation->get_slice();
+        uint8_input_operands.insert(slice.input_operand_id);
+        break;
+      }
+      case mojom::Operation::Tag::kSplit: {
+        const auto& split = *operation->get_split();
+        uint8_input_operands.insert(split.input_operand_id);
+        break;
+      }
+      case mojom::Operation::Tag::kTile: {
+        const auto& tile = *operation->get_tile();
+        uint8_input_operands.insert(tile.input_operand_id);
+        break;
+      }
+      case mojom::Operation::Tag::kTranspose: {
+        const auto& transpose = *operation->get_transpose();
+        uint8_input_operands.insert(transpose.input_operand_id);
+        break;
+      }
+      case mojom::Operation::Tag::kTriangular: {
+        const auto& triangular = *operation->get_triangular();
+        uint8_input_operands.insert(triangular.input_operand_id);
+        break;
+      }
+      case mojom::Operation::Tag::kWhere: {
+        const auto& where = *operation->get_where();
+        uint8_input_operands.insert(where.true_value_operand_id);
+        uint8_input_operands.insert(where.false_value_operand_id);
+        break;
+      }
+      default:
+        break;
+    }
+  }
+
+  // The output operands can also be viewed as the uint8 input operands.
+  for (const auto& output_operand_id : graph_info_->output_operands) {
+    uint8_input_operands.insert(output_operand_id);
   }
 
-  // Find all operands that can skip cast operators. If an operand is a bool
-  // output and also used as bool input, it can skip inserting cast operator
-  // to/from uint8.
+  // Find all operands that can skip cast operators. If a bool output operand is
+  // used as bool input and not used as uint8 input at the same time, it can
+  // skip inserting cast operator to/from uint8.
   for (auto id : bool_output_operands) {
-    if (bool_input_operands.count(id)) {
+    if (bool_input_operands.count(id) && !uint8_input_operands.count(id)) {
```

Comment:
Could we keep the previous `bool_operands_` and add `uint8_input_operands_`?

When adding a node that has bool output, such as `lesser`, if its output is used by a node that requires a uint8 input (in `uint8_input_operands`), such as `add`, appending a `cast`.

When adding a node that has bool input, such as 'where', if its input is an bool operand (e.g. in `bool_operands_`), use it directly without prepending a `cast`.

Would it lead us to the desired solution?
```
Less
  |  \
  |   Cast 
  |     |
Where  Add  
```

Reply:
Thanks for your suggestion from another perspective, which led us to the desired solution!
Now we introduced `bool_to_uint8_operands_` set to store operands that should be inserted cast operators from bool to uint8 data type. And `introduced bool_to_uint8_operands_` map to map operand and bool operand. Please take another look, thanks a lot!